README.md
deploy-helper Cookbook
======================
It leaves the application deployment purely in the
hands of the developer.  It provides a clean way to
gather deployment details from an application repo
that can then be used during the primary application
deployment.

This is very basic and only supports git at the moment.

Usage
-----
Add a .deploy.yml file to the git application repo
you plan to use for an application deployment and
use the helper method.

```yaml
---
revision:
  default: master
  production: prod
  staging: staging
  dev: dev
migrate:
  default: false
  proudction: true
```

The helper method returns a mash of the yaml info,
that should represent the Chef environment so if
revision is set for production but not master it
will just use whatever the default is.

```ruby
app_repo = 'git@github.com:jarosser06/magic'
deploy_key = data_bag_item('secrets', 'deploy_key')
app_info = git_deployment_info(app_repo, deploy_key['key'])
```

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
- Author:: Jim Rosser(jarosser06@gmail.com)

```text
copyright (C) 2015 Jim Rosser

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without restriction,
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software,
and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
```


CHANGELOG.md
deploy-helper CHANGELOG
=======================

0.0.2
-----
- [jarosser06] - Removed random puts of deploy key


0.0.1
-----
- [jarosser06] - Initial release of deploy-helper

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


test/cookbooks/app_deploy-test/metadata.rb
name             'app_deploy-test'
maintainer       'Jim Rosser'
maintainer_email 'jarosser06@gmail.com'
license          'mit'
description      'Tests the app_deploy Resources'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '0.1.0'
depends          'deploy-helper'


test/cookbooks/app_deploy-test/recipes/magic.rb
app_repo = 'git@github.com:jarosser06/magic'
deploy_key = data_bag_item('secrets', 'deploy_key')
info = git_deployment_info(app_repo, deploy_key['key'])

file '/root/id_deploy' do
  action :create
  mode 0400
  content deploy_key['key']
end

template '/root/ssh_wrapper' do
  action :create
  mode 0554
  source 'git-ssh-wrapper.erb'
  cookbook 'deploy-helper'
  variables(cache_dir: '/root')
end

Chef::Log.warn(info)

deploy_revision 'myapp' do
  repo app_repo
  revision info['revision'] || 'master'
  ssh_wrapper '/root/ssh_wrapper'
  symlink_before_migrate.clear
  symlinks.clear
  migrate info['migrate'] || false
  migration_command info['migration_command']
end


libraries/deploy_helper.rb
require 'yaml'
require 'chef/mash'
require 'net/http'
require 'chef/recipe'
require 'chef/resource/directory'
require 'chef/provider/directory'

module Application
  module DeployHelper

    def git_deployment_info(repo, deploy_key=nil, revision='master')
      install_git
      app = Application.new(run_context, repo, deploy_key, revision)
      app.info(node.chef_environment)
    end

    def install_git
      pkg = Chef::Resource::Package.new('git', run_context)
      pkg.run_action :install
    end

    class Application
      def initialize(run_context, repo, deploy_key=nil, revision='master')
        @run_context = run_context
        @repo = repo
        @deploy_key = deploy_key
        @revision = revision
        @cache_dir = ::File.join(Chef::Config[:file_cache_path],
                                 repo_name)
      end

      def info(environment)
        create_cache_dir

        unless @deploy_key.nil?
          create_ssh_wrapper
          create_deploy_key
        end
        checkout_git_repo

        deploy_yaml = ::File.join(@cache_dir, 'cache/.deploy.yml')
        return nil unless ::File.exist? deploy_yaml
        yaml = YAML.load(IO.read(deploy_yaml))

        compiled_info = Mash.new
        yaml.each do |k, v|
          if v.has_key? environment
            compiled_info[k] = v[environment]
          else
            compiled_info[k] = v['default']
          end
        end
        compiled_info
      end

      def checkout_git_repo
        repo = Chef::Resource::Git.new(repo_name, @run_context)
        repo.repository @repo
        repo.ssh_wrapper ssh_wrapper unless @deploy_key.nil?
        repo.reference @revision
        repo.destination ::File.join(@cache_dir, 'cache')
        repo.run_action :sync
      end

      def create_cache_dir
        dir = Chef::Resource::Directory.new(@cache_dir, @run_context)
        dir.mode 0755
        dir.owner 'root'
        dir.group 'root'
        dir.run_action :create
      end

      def repo_name
        return @repo_name unless @repo_name.nil?
        begin
          uri_path = URI.parse(@repo).path
          @repo_name = uri_path.byteslice(1, uri_path.length)
        rescue
          @repo_name = @repo.split(':').last
        end
        @repo_name.gsub!('/', '_')
        @repo_name
      end

      def ssh_wrapper
        return @ssh_wrapper unless @ssh_wrapper.nil?
        @ssh_wrapper = ::File.join(@cache_dir, '.ssh_wrapper')
        @ssh_wrapper
      end

      def create_ssh_wrapper
        script = Chef::Resource::Template.new(ssh_wrapper, @run_context)
        script.mode 0554
        script.owner 'root'
        script.group 'root'
        script.source 'git-ssh-wrapper.erb'
        script.cookbook 'deploy-helper'
        script.variables(cache_dir: @cache_dir)
        script.run_action :create
      end

      def create_deploy_key
        key = ::File.join(@cache_dir, 'id_deploy')
        deploy_key = Chef::Resource::File.new(key, @run_context)
        deploy_key.mode 0400
        deploy_key.sensitive true
        deploy_key.owner 'root'
        deploy_key.group 'root'
        deploy_key.content @deploy_key
        deploy_key.run_action :create
      end
    end
  end
end

Chef::Recipe.send(:include, ::Application::DeployHelper)
Chef::Resource.send(:include, ::Application::DeployHelper)
Chef::Provider.send(:include, ::Application::DeployHelper)
